language-support/ts/daml-ledger/index.ts

```diff
@@ -487,28 +495,39 @@ class Ledger {
     template: Template<T, K, I>,
     endpoint: string,
     request: unknown,
+    reconnectRequest: () => unknown,
     init: State,
     change: (state: State, events: readonly Event<T, K, I>[]) => State,
   ): Stream<T, K, I, State> {
     const protocols = ['jwt.token.' + this.token, 'daml.ws.auth'];
-    const ws = new WebSocket(this.wsBaseUrl + endpoint, protocols);
-    let isLive = false;
+    let ws = new WebSocket(this.wsBaseUrl + endpoint, protocols);
+    let isLiveSince: undefined | number = undefined;
+    let lastOffset: undefined | string = undefined;
     let state = init;
+    let isReconnecting: boolean = false;
     const emitter = new EventEmitter();
-    ws.addEventListener('open', () => {
-      ws.send(JSON.stringify(request));
-    });
-    ws.addEventListener('message', event => {
+    const onOpen = (): void => {
+      if (isReconnecting) {
+          ws.send(JSON.stringify({ 'offset': lastOffset }));
+          ws.send(JSON.stringify(reconnectRequest()));
+      } else
+        ws.send(JSON.stringify(request));
+    };
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const onMessage = (event: { data: any; type: string; target: WebSocket } ): void => {
       const json: unknown = JSON.parse(event.data.toString());
       if (isRecordWith('events', json)) {
         const events = jtv.Result.withException(jtv.array(decodeEvent(template)).run(json.events));
         if (events.length > 0) {
           state = change(state, events);
           emitter.emit('change', state, events);
         }
-        if (isRecordWith('offset', json) && !isLive) {
-          isLive = true;
-          emitter.emit('live', state);
+        if (isRecordWith('offset', json)) {
+          lastOffset = jtv.Result.withException(jtv.string().run(json.offset));
+          if (!isLiveSince) {
+            isLiveSince = new Date().getTime();
+            if (!isReconnecting) emitter.emit('live', state);
+          }
```

Comment:
Let's be more explicit:
```suggestion
          if (isLiveSince === undefined) {
            isLiveSince = Date.now();
            if (!isReconnecting) {
              emitter.emit('live', state);
            }
          }
```
`Date.now()` should appear somewhere further down too.